### PR TITLE
Rewrite bash scripts into go

### PR DESCRIPTION
**Pull request by dergeberl**
Open Tasks:
- [x] add docs for local setup and testing
- [x] check README.md
- [x] reduce log from `pathcontroller`
- [x] remove shell scripts 🎉 
- [x] check memory usage - improved for VPNShoot
- [x] check container images (before the PR there was less other software, like `ls`)
  - [x] check if it is possible to have a dev image with useful tools

**What this PR does / why we need it**:

This PR rewrites the bash scripts into go. This will made it easier to adopt to new features in the future.
- Some ENVs are different or removed, this needs to be rolledout together with the changes in the gardener/gardener https://github.com/gardener/gardener/pull/9774
- The `initContainers` use a subcommand
- Everything that was done with `iproute2` is now done with the `netlink` package
- Merge `Dockerfile`s into one `Dockerfile`

Co-authored-by: @MartinWeindel 
Co-authored-by: @hown3d
Co-authored-by: @mreiger

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
⚠️ The pipeline jobs for building images are failing as changes in `.ci/pipeline_definitions` are not yet used. They become only active after merging to master branch.
 
⚠️ This is a breaking change and needs also changes to gardener/gardener

**Release note**:

```breaking operator
Rewrite bash scripts into go. 
Log output changed. Expected ENVs from the containers changed. More memory is needed because of the change to go.
```

### CODE REVIEW

**Comment by gardener-robot-ci-3** [comment]
Thank you @dergeberl for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

**Comment by MartinWeindel** [comment]
Regarding memory usage these are my major findings:
- Investigating the heap usage of the Go executables found only very small heap size about 1-2 MB. So this not the cause.
- The RSS (Resident Set Size) was still large if the openvpn was run in a separate container. There seems to be an memory overhead caused by Rosetta on Macbooks with M1/M2 CPUs when using local Gardener setup
- Memory consumption drops significantly if local Gardener setup run directly on Linux machine
- Remaining memory consumption seems to be related to pages of the go executable itself
- After moving the start/restart loop back into a minimal shell script (commit https://github.com/gardener/vpn2/pull/84/commits/09bf6f6f7810edacd32773586cffbd654c0edf6f), memory usage was back to same level as before the rewrite.

```
vpn-seed-server-665cf779b9-52vwb             3m           19Mi
---
vpn-shoot-f956748d-gv5h5                     5m           1Mi
```

Without the script, on Linux the memory consumption was ~ 6Mi

File: cmd/seed_server/app/firewall.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	cmd.MarkFlagsRequiredTogether("device", "mode")
+
+	return cmd
+}
+
+func runFirewallCommand(log logr.Logger, device, mode string) error {
+	// Firewall subcommand is called indirectly from openvpn. As PATH env variables seems not to be set,
+	// it is injected here.
+	os.Setenv("PATH", "/sbin")
+	iptable4, err := iptables.New(iptables.IPFamily(iptables.ProtocolIPv4))
```
Comment (axel7born) [inline cmd/seed_server/app/firewall.go:49]: We added an iptables backend detection with https://github.com/gardener/vpn2/pull/63.
We have seen issues on some os versions, that not all required modules are present in the vpn-seed-server container.
How is this handled in `github.com/coreos/go-iptables/iptables`? Is there a similar check to determine the iptables backend?

Comment (MartinWeindel) [reply cmd/seed_server/app/firewall.go:49]: Looking into the code of `github.com/coreos/go-iptables/iptables`, it always uses either `iptables` or `ip6tables` command. It has a detection for nft or legacy mode, but it is not used to set the iptables executable name.

@axel7born
Can you explain how to reproduce the issue you fixed with #63 ?

Comment (axel7born) [reply cmd/seed_server/app/firewall.go:49]: We have seen the issue with gardenlinux 1312.0.0.  There the default switched from iptables-legacy to iptables-nft.
The vpn container still used iptables-legacy and for the ipv6 case not all required kernel modules were present. I can't reproduce the isssue with newer gardenlinux versions. 
We have also seen the issue in gke soil clusters. But also there, I can't reproduce it now.
To reproduce the issue, exec into a vpn pod, and try to run the commands `ip6tables-legacy -L` and `ip6tables-nft -L`. When the needed kernel modules are not present, the command returns an error.
When we just run `iptables` there is a risk that the command doesn't work due to missing kernel modules.

Comment (MartinWeindel) [reply cmd/seed_server/app/firewall.go]: Added iptables backend detection, see below

File: pkg/utils/iptables.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+// NewIPTables wraps the creation of IPTables to patch the path to the correct implementation binary.
+// It has been introduced to avoid the risk that the command doesn't work due to missing kernel modules.
+func NewIPTables(log logr.Logger, proto iptables.Protocol) (*iptables.IPTables, error) {
+	t, err := iptables.New()
+	if err != nil {
+		return nil, err
+	}
+
+	t2 := (*ipTables)(unsafe.Pointer(t))
```
Comment (hown3d) [inline pkg/utils/iptables.go:39]: I suggest doing a PATH lookup for iptables-legacy and iptables-nft, executing the commands to check which backend is working and creating iptables.New() with iptables.Path() option, so we don't need to do this for private field access.

Comment (MartinWeindel) [reply pkg/utils/iptables.go:39]: Thanks for the hint! The `Path()` method is only available on the master branch, therefore I missed it.

File: pkg/openvpn/config_server.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+script-security 2
+up "/bin/seed-server firewall --mode up --device {{ .Device }}"
+down "/bin/seed-server firewall --mode down --device {{ .Device }}"
+
+{{ if not (eq .StatusPath "") -}}
+status {{ .StatusPath }} 15
+status-version 2
+{{- end -}}
+`
+
```
Comment (axel7born) [inline pkg/openvpn/config_server.go:81]: Should we maybe put the templates in separate files? WDYT?

Comment (MartinWeindel) [reply pkg/openvpn/config_server.go]: Moved templates to own files

File: pkg/utils/iptables.go
Comment (hown3d) [inline pkg/utils/iptables.go:1]: Can we move this into pkg/network/iptables?

utils could get crowded pretty fast, utils in general should be avoided

File: pkg/utils/iptables.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+	"github.com/coreos/go-iptables/iptables"
+	"github.com/go-logr/logr"
+)
+
+// NewIPTables wraps the creation of IPTables to patch the path to the correct implementation binary.
+// It has been introduced to avoid the risk that the command doesn't work due to missing kernel modules.
+func NewIPTables(log logr.Logger, proto iptables.Protocol) (*iptables.IPTables, error) {
+	for _, suffix := range []string{"legacy", "nft"} {
+		if path := "/sbin/iptables-" + suffix; iptablesWorks(path) {
```
Comment (hown3d) [inline pkg/utils/iptables.go:20]: Rather use exec.LookPath instead of hardcoding the path.

**Review by hown3d — CHANGES_REQUESTED** [review]

File: pkg/utils/iptables.go
Comment (MartinWeindel) [reply pkg/utils/iptables.go:1]: moved

File: pkg/utils/iptables.go (hunk 4 above)
Comment (MartinWeindel) [reply pkg/utils/iptables.go]: I prefer to keep the hardcoded path, as it is well-defined by the alpine package.

**Comment by axel7born** [comment]
With ha vpn we have the shoot-client container running in the seed in the kube-apiserver pod. This is already confusing. The path-controller is only needed in seed. Maybe, we should we use the opportunity to separate seed-client and shoot client? Or change the name from `shoot-client` to `vpn-client`?

File: cmd/shoot_client/app/pathcontroller/ping.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	err = c.SetReadDeadline(deadline)
+	if err != nil {
+		return fmt.Errorf("error setting deadline: %w", err)
+
+	}
+
+	msg := icmp.Message{
+		Type: ipv4.ICMPTypeEcho, Code: 0,
+		Body: &icmp.Echo{
+			ID: os.Getpid() & 0xffff, Seq: 1,
```
[the remark this answers is not in the document]
Comment (MartinWeindel) [reply cmd/shoot_client/app/pathcontroller/ping.go:61]: The code follows the example given at [pkg.go.dev/golang.org/x/net/icmp#PacketConn](https://pkg.go.dev/golang.org/x/net/icmp#PacketConn) (look for Example (NonPrivilegedPing)).
To create the message on the fly looks good to me. I would prefer to keep it simple as there is no need for optimisation (it is called twice every 2 seconds).

**Comment by MartinWeindel** [comment]
> With ha vpn we have the shoot-client container running in the seed in the kube-apiserver pod. This is already confusing. The path-controller is only needed in seed. Maybe, we should we use the opportunity to separate seed-client and shoot client? Or change the name from `shoot-client` to `vpn-client`?

As we need to rename the images anyway to be able to switch back to the old implementation, I propose to use `vpn-client` instead of `shoot-client` for package and image name and similar `vpn-server` instead of `seed-server`.

I don't see a big win in separating "seed-client" and "shoot-client", as they share many things.

**Review by MartinWeindel — APPROVED** [review]

**Pull request merged**